Embedded YarnEditor webview custom editor

TranquilMarmot opened the pull request:
# What is this?

Hello friends! :wave:  Thank you for Yarn Spinner; this tool is awesome! :yarn: :cat2: 

A while back I saw [this draw.io extension](https://github.com/hediet/vscode-drawio) for VSCode and was blown away by the idea of embedding a full JS app inside of an editor... so naturally when I came across the [Yarn Editor](https://yarnspinnertool.github.io/YarnEditor/), the first thing I thought was... "how long would it take to embed this editor in VSCode?!" The answer is... about the length of a long weekend.

This uses a [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) to include the `YarnEditor` repo. Currently, this PR has the submodule pointed to [my fork/branch of YarnEditor](https://github.com/TranquilMarmot/YarnEditor/commits/tranquilmarmot/vscode-yarn-spinner). There is a PR against `YarnEditor` that is needed for this to work: https://github.com/YarnSpinnerTool/YarnEditor/pull/186

Once that PR is merged, this one can be updated to have the proper submodule that points to the _real_ `YarnEditor`.

This is in TypeScript. If you've never encountered TypeScript before don't fret! This PR doesn't do anything too crazy with it.

This has been tested on both Windows and Linux.

Also, don't freak out at the `+2,000` lines added! Most of that is `package-lock.json`.

For reference on how a lot of this was done, check out:
https://code.visualstudio.com/api/extension-guides/webview
https://code.visualstudio.com/api/extension-guides/custom-editors
https://github.com/microsoft/vscode-extension-samples/tree/master/custom-editor-sample

## Features

- Open, edit, and save `.yarn` and `.yarn.txt` files in your workspace in the Yarn Editor without leaving VSCode (this is killer for Unity projects!!!)
- Saving is handled by VSCode and it will mark files as dirty when they're changed (so you don't accidentally close them after making changes!)
- You can also run the "vanilla" editor with a "Start Yarn Editor" command via the command palette (<kbd>F1</kbd> menu). This is basically the equivalent of the Electon app... but in a VSCode extension!
- Options are configured in VSCode's settings instead of in the editor. When running the "vanilla" editor, it will still show the "File" menu and let you change settings, but they will not persist.
- Automatically detect when the user has a "dark" VSCode theme and switch to night view. This can be overridden by a special option in the VSCode extension settings.
- Easily swap between the editor and the syntax-highlighted text view
- When viewing a diff, VSCode will open two editors side-by-side with the old and new files

## Screenshots

(Note: these were all taken before a bug with arrow rendering position was fixed)

_`Sally.yarn` sample, with automatically detected night mode:_
![vscode-yarn-sally](https://user-images.githubusercontent.com/504259/82779577-65eac380-9e09-11ea-9782-454882a5252d.png)

_`Sally.yarn` sample, with blueprint theme:_
![vscode-sally-blueprint](https://user-images.githubusercontent.com/504259/82779635-8b77cd00-9e09-11ea-8a00-85f44d2d25a2.png)

_VSCode settings for the extension; these map 1:1 to the settings modal and even have the same dropdown options:_
![vscode-yarn-settings](https://user-images.githubusercontent.com/504259/82779586-6be0a480-9e09-11ea-9171-850181890511.png)

_Switching between the syntax-highlighted text view and the editor view. This can be done by clicking the `...` menu button on the VSCode tab and choosing "Reopen With..."_
![vscode-yarn-switch](https://user-images.githubusercontent.com/504259/82780364-c24ee280-9e0b-11ea-9bbf-2bd7cbfde24e.png)

_`StressTest.yarn` open alongside another file; stress test opens surprisingly quickly!_
![vscode-yarn-stress](https://user-images.githubusercontent.com/504259/82779597-6f742b80-9e09-11ea-8e83-7d8f389d13c4.png)

## How to try this out

Checkout this branch and then run `git submodule init && git submodule update`. This _should_ pull down the proper branch of `YarnEditor`.

Then, run `npm install && npm run yarneditor:build`. This will build `YarnEditor` and copy it to `out/dist` (the extension's code is transpiled to JS and copied to `out` as well when it runs).

You can then hit <kbd>F5</kbd> to run the extension. It will open a new VSCode window with the extension loaded. From there, try opening a `.yarn` file or hitting <kbd>F1</kbd> to run the "Start Yarn Editor" command.

## Notes

The amount of HTML string manipulation in here doesn't feel great, but honestly it's not all that bad and YarnEditor's layout and HTML made it easy.

### Icons

Originally, I had this extension include an "icon set" but it seems like VSCode can only have one icon set active at a time right now. If we want the yarn icon to properly appear for `.yarn` files it seems like a PR would need to be opened against https://github.com/hellopao/vscode-seti-icons

This issue against VSCode tracks the ability to have extension provide just one icon: https://github.com/microsoft/vscode/issues/14662

### Undo/redo

I couldn't manage to get native undo/redo working for changes. It seems like the editor would need to be reworked to only track specific lines in the underlying file changing to get this to work properly. Undoing/redoing adding/removing nodes still works within the editor, though.

### Lint

Okay, for some people lint is a touchy subject. Currently, this is just using the most basic lint rules that work with TypeScript and [Prettier](https://prettier.io/).

If you want, we can make this obey the [same lint rules as YarnEditor](https://github.com/YarnSpinnerTool/YarnEditor/blob/master/.eslintrc.js) and we can add a `prettierrc.json` to make Prettier do the right things.

TranquilMarmot commented on .gitmodules:
```diff
@@ -0,0 +1,4 @@
+[submodule "YarnEditor"]
+	path = YarnEditor
+	url = https://github.com/TranquilMarmot/YarnEditor.git
```
This will need to be updated once https://github.com/YarnSpinnerTool/YarnEditor/pull/186 is merged

TranquilMarmot commented on package.json:
```diff
@@ -1,49 +1,240 @@
+    "theme": "light"
+  },
+  "engines": {
+    "vscode": "^1.44.0"
```
This had to be bumped up from `1.38.0` since `1.44.0` was the most recent version to fully support custom editors (this is some cutting edge stuff! :sunglasses: )

TranquilMarmot commented on package.json:
```diff
@@ -1,49 +1,240 @@
+        "displayName": "Yarn Editor",
+        "selector": [
+          {
+            "filenamePattern": "*.yarn"
```
I considered also adding `.json`, `.twee`, and all the other supported formats with `"priority": "optional"` here but I'm not sure how supported those are in the editor.

TranquilMarmot commented on package.json:
```diff
@@ -1,49 +1,240 @@
+        "path": "./yarn-icon-theme.json"
+      }
+    ],
+    "configuration": {
```
All of these configuration settings were ripped straight from `settings.js`. This makes it so that you can change settings in your VSCode preferences!

daviddq commented:
Super cool! 👏👏👏👏👏👏👏

TranquilMarmot commented:
Okay, I think this is finished! Got everything working smoothly, including settings and marking files as dirty and letting VSCode handle saving/loading.

Here's a copy of the extension built with [`vsce`](https://code.visualstudio.com/api/working-with-extensions/publishing-extension):
[yarn-spinner-1.2.0-2b4e5bf.zip](https://github.com/YarnSpinnerTool/VSCodeExtension/files/4679554/yarn-spinner-1.2.0-2b4e5bf.zip)

GitHub wouldn't let me upload it as a `.vsix` file, so to install it:
- Download and change the extension to `.vsix` (this _is_ just a `.zip` file, so if you want you can open it and look at what's inside first)
- Run `code --install-extension yarn-spinner-1.2.0-2b4e5bf.vsix`
- Restart VSCode if you have it running
- Open a `.yarn` file!

blurymind commented:
Thanks for working on this! It does look very cool 👍👌
I will try to do a review and get it merged asap

blurymind commented:
correct me if I am wrong, but once merged, this will become available via
https://marketplace.visualstudio.com/items?itemName=SecretLab.yarn-spinner

I will review the changes required on the yarnEditor side of things. Somebody else needs to review it on the yarnSpinner side :)

I am pretty excited to see this in vscode. Very very cool!

blurymind commented:
> Okay, I think this is finished! Got everything working smoothly, including settings and marking files as dirty and letting VSCode handle saving/loading.
> 
> Here's a copy of the extension built with [`vsce`](https://code.visualstudio.com/api/working-with-extensions/publishing-extension):
> [yarn-spinner-1.2.0-2b4e5bf.zip](https://github.com/YarnSpinnerTool/VSCodeExtension/files/4679554/yarn-spinner-1.2.0-2b4e5bf.zip)
> 
> GitHub wouldn't let me upload it as a `.vsix` file, so to install it:
> 
> * Download and change the extension to `.vsix` (this _is_ just a `.zip` file, so if you want you can open it and look at what's inside first)
> * Run `code --install-extension yarn-spinner-1.2.0-2b4e5bf.vsix`
> * Restart VSCode if you have it running
> * Open a `.yarn` file!

following these instructions on Linux, I get
```
code --install-extension yarn-spinner-1.2.0-2b4e5bf.vsix
Installing extensions...
Unable to install extension 'secretlab.yarn-spinner' as it is not compatible with VS Code '1.43.0'.
Failed Installing Extensions: yarn-spinner-1.2.0-2b4e5bf.vsix
```

however installing the official extension from market place works fine

blurymind commented:
Ok I tested this and found some problems on my end. Not sure if they need addressing at this pr or the other. Noted them here
https://github.com/YarnSpinnerTool/YarnEditor/pull/186#issuecomment-634651197

blurymind commented:
There are a bunch of recent regressions in yarnEditor that need urgent fixing. Will this extension always use the latest version of yarnEditor? If not, can we update it to use a more recent version - after we fix the recent regressions?

TranquilMarmot commented:
Once this is merged, then the next version of https://marketplace.visualstudio.com/items?itemName=SecretLab.yarn-spinner released will have it, yes. The editor and the syntax highlighting could be split out into separate extensions, but as it is it's really easy to swap your default editor in VSCode from the yarn editor back to syntax-highlighted text.

This extension will not always have the latest version of yarn editor, but since it's using a [git submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) it will be easy to update once the `.gitmodules` file points to https://github.com/YarnSpinnerTool/YarnEditor/ Updating it should be as easy as `git submodule --remote` then doing another release of the extension. I'll add that to the README.

blurymind commented:
the urgent fixes to yarnEditor were merged yesterday:) can you update the extension to use the latest master please :)

desplesda commented:
@TranquilMarmot, this is absolutely astonishing. Well done!

Currently, the Yarn Editor view presents the text editor as a nested view inside the editor. Something that I'd like to consider is the ability to open nodes in the built-in text editor. (@blurymind and I had a conversation a few months back about replacing the Ace editor currently shipping in Yarn Editor with Monaco; this could be a way to do that.)

Does VSCode have a concept like "sub-documents", where the user would double-click on a node and get presented with a text editor that's a limited subset of the document? If that isn't possible, maybe it could (re-)open the document and jump to the line where the node begins?

TranquilMarmot commented:
@blurymind okay, https://github.com/YarnSpinnerTool/YarnEditor/pull/186 and this branch have both been updated to be at the latest version of YarnEditor.

Here's a new build of the extension (again, rename to `.vsix` to install): [yarn-spinner-1.2.0-0b6555c.zip](https://github.com/YarnSpinnerTool/VSCodeExtension/files/4699638/yarn-spinner-1.2.0-0b6555c.zip)

@desplesda that's so funny; as I was doing this, I couldn't help but think "it would be nice if this had monaco as the editor..."

It looks like VSCode does have [Virtual Documents](https://code.visualstudio.com/api/extension-guides/virtual-documents) but right now they're read-only. There is an open issue to make these editable: https://github.com/microsoft/vscode/issues/10547.

It may seem silly since we'd already be in VSCode, but maybe it's not too crazy to include monaco in the extension anyway. Another thing to think about: having it open nodes inside a VSCode text window means the editing toolbar would no longer be there which means no more buttons for preview, add image, add command, etc.

There is also the concept of [File System Providers](https://github.com/microsoft/vscode-extension-samples/tree/master/fsprovider-sample) that might be able to be leveraged, but it would be pretty weird. You could treat every node in the yarn file as a separate virtual file that opens when you double-click on a node. This might require parsing the Yarn file in JS but it sounds like from what @blurymind stated above, bondage might need some updates for something like that to happen? Alternatively, the editor webview could send back all the nodes to the extension when it's done loading but that probably wouldn't be too efficient.

TranquilMarmot commented:
Another approach I just tried out; when a node is opened...
- Create a temporary file on-disk with a `.yarnNode` extension (which is also registered to have the syntax highlighting; the different extension is so it doesn't trigger the custom editor)
- Write the node contents to it
- Use node.js to watch the file
- Whenever the file changes, send a message to the app telling it to update that node (by name)
- When the extension is "deactivated" (i.e. the window is closed) delete all the temporary files

This actually ends up working really well!!! Only issue is you can no longer rename nodes. I think I'm going to add an "Open in Visual Studio Code Editor" button to the node editor box.

blurymind commented:
@desplesda  swapping ace with monaco in yarnEditor is no small job and should be done in some other pr. Whoever does that PR will need to also write a mode for monaco to give yarn the correct syntax colouring, autocompletion and everything else that currently works with ace editor

TranquilMarmot commented:
Alright, here's what I ended up with RE: Editing nodes in VSCode's text editor!

By default, nodes still open in ace but there's now a button to edit them in VSCode (this doesn't show up in the web app; I checked 😉 )
![image](https://user-images.githubusercontent.com/504259/83317591-b041bb00-a1e2-11ea-9c22-8ac780239a0a.png)

Clicking this button will create a new, temporary file and open it up. At the same time, it adds a node.js file watcher to it that  listens for changes; when the file changes, the extension reads it and sends it back to the editor which will then update the node. This lets you edit node titles, tags, and bodies (color and position can still be changed from the yarn editor):
![image](https://user-images.githubusercontent.com/504259/83317764-ecc1e680-a1e3-11ea-98dc-18516e4be053.png)

This ends up working surprisingly well! If you open up two nodes, you actually get two temporary files; the extension is listening to both of them and changes made to either are sent back to the editor. This means you can do cool things like editing two nodes side-by-side!
![image](https://user-images.githubusercontent.com/504259/83317838-b6389b80-a1e4-11ea-9578-32b92ae6aaf2.png)

The extension will automatically delete any temporary files when it "deactivates" (i.e. when the window is closed).

On top of that, I added an "Always Open Nodes In Visual Studio Code Editor" option that will make it so that double-clicking on a node will _always_ open it in the VSCode text editor.

![image](https://user-images.githubusercontent.com/504259/83317908-5393cf80-a1e5-11ea-93ad-dbf61a74eb61.png)

TranquilMarmot commented:
Okay, fixed a slew of bugs with the "Edit in Visual Studio Code Text Editor" functionality and updated the `YarnEditor` branch to be in line with `master`.

Another `.vsix`: [yarn-spinner-1.2.0-e711fbb.zip](https://github.com/YarnSpinnerTool/VSCodeExtension/files/4708678/yarn-spinner-1.2.0-e711fbb.zip)

And another really cool screenshot of editing two nodes at once 😄 
![image](https://user-images.githubusercontent.com/504259/83365444-fa4eac00-a35c-11ea-8315-c360aeb1edba.png)
As soon as you save one of the temporary files, the updates to the `.yarn` file are pretty much instant! It feels great.

desplesda commented:
What's needed before we can merge this PR?

blurymind commented:
a bit more testing - see if you find things not working as expected :)
Also the other (yarnEditor) pr to get merged

TranquilMarmot commented:
Updated this PR and the other one to have the latest `YarnEditor`; here's another `.vsix`: [yarn-spinner-1.2.0-aa30f93.zip](https://github.com/YarnSpinnerTool/VSCodeExtension/files/4739203/yarn-spinner-1.2.0-aa30f93.zip)

@desplesda I'd say this is done. Would definitely love folks to do some more testing! Once https://github.com/YarnSpinnerTool/YarnEditor/pull/186 is merged I can update this PR to have its submodule point to the right repo.

blurymind commented:
The changes required on the yarnEditor side for this to work have now been marged to the master branch
:+1:

TranquilMarmot commented:
Okay! The submodule in this has been update to point to the real YarnEditor which has the needed changes merged in. This should also be good to go now.

daviddq commented:
What's missing for an official release of the extension? I'm excited to get the latest fixes @TranquilMarmot 's  😄

blurymind commented:
I think this can be merged now, but waiting for someone on this repo to review and merge it :)

desplesda commented:
I'll review this PR early this coming week! Excited to get this merged.

TranquilMarmot commented:
Hey all, as good as this PR works with Yarn Editor, there were a few things that didn't quite sit right with me:

- No shared code between the editor and the extension; the extension ends up re-implementing a lot of logic from the editor
- The whole git submodule thing is quite painful to deal with (this PR is already horribly out of date 😅  )
- The state management between the editor and the extension isn't great
   - Currently, Yarn Editor has to send all nodes to the extension with every change
   - The main issue with this is that it breaks down VSCode's concept of changes and its undo/redo functionality
- The hacks that had to be done to get the editor to embed in the VSCode webview are pretty nasty and are brittle. They are likely to break with future changes to Yarn Editor.
- Ace editor is a large dependency that is not needed in VSCode
- The Yarn Editor embedded in VSCode has stylistic clashes (colors, icons, overall design, etc.)
- No tests in the current Yarn Editor made it difficult to know if I broke anything when making changes to it 🙈 

---

All of the above inspired me to spend the past month putting together YarnLoom, a more native and tightly-integrated version of the Yarn Editor that plays nicely with VSCode:

https://github.com/tranquilmarmot/yarnloom
https://marketplace.visualstudio.com/items?itemName=TranquilMarmot.yarn-spinner-loom

Any `.yarn` files made with Yarn Editor work with YarnLoom and vice versa! It uses the same mechanisms for node colors, tags, and placement.

Key differences:
- Common library of functions used for parsing/generating yarn files that are shared between the editor and the extension
- The editor sends much more specific messages back to the extension so that it can track changes for better undo/redo functionality
   - This means node renames, additions, color changes, edits, and everything else support native undo/redo
   - See the YarnLoom [CONTRIBUTING.md](https://github.com/TranquilMarmot/YarnLoom/blob/main/CONTRIBUTING.md#communication-between-loom-extension-and-loom-editor) doc for more info
- Uses VSCode supplied CSS variables for colors so that it matches the user's VSCode theme
- Uses [codicons](https://github.com/microsoft/vscode-codicons) for icons so they match VSCode
- Additional features
   - More tag management tools
   - Renaming nodes also renames links to those nodes
   - Option for case sensitivity and regex when searching
   - Better feedback around renaming/adding nodes with names that already exist
   - Uses D3 for drawing node relationships, resulting in much smoother rendering
- Missing features (these may be added later)
   - Preview using bondage.js
   - Selecting and manipulating multiple nodes at once
   - BBCode support
   - JSON/XML/twee/twee2 import/export support (only `.yarn` files are supported)
- Lots and lots of tests 😄 
- 100% TypeScript 😉 

---

You're welcome to merge this and keep it!

I'll be happy to answer any questions you have but will probably be putting more effort into maintaining YarnLoom.

blurymind commented:
Beautiful work on this. I love that you used typescript and react for it.

I assume its not yet fully on par in features but really love what you did
with aspects of the design. I was thinking of doing the same for the node
colour picker.

It would be amazing if loom can function outside of vscode as a pwa and an
electron app too

On Sat, 1 Aug 2020, 21:08 Nate Moore, <notifications@github.com> wrote:

> Hey all, as good as this PR works with Yarn Editor, there were a few
> things that didn't quite sit right with me:
>
>    - No shared code between the editor and the extension; the extension
>    ends up re-implementing a lot of logic from the editor
>    - The whole git submodule thing is quite painful to deal with (this PR
>    is already horribly out of date 😅 )
>    - The state management between the editor and the extension isn't
>    great
>       - Currently, Yarn Editor has to send all nodes to the extension
>       with every change
>       - The main issue with this is that it breaks down VSCode's concept
>       of changes and its undo/redo functionality
>    - The hacks that had to be done to get the editor to embed in the
>    VSCode webview are pretty nasty and are brittle. They are likely to break
>    with future changes to Yarn Editor.
>    - Ace editor is a large dependency that is not needed in VSCode
>    - The Yarn Editor embedded in VSCode has stylistic clashes (colors,
>    icons, overall design, etc.)
>    - No tests in the current Yarn Editor made it difficult to know if I
>    broke anything when making changes to it 🙈
>
> ------------------------------
>
> All of the above inspired me to spend the past month putting together
> YarnLoom, a more native and tightly-integrated version of the Yarn Editor
> that plays nicely with VSCode:
>
> https://github.com/tranquilmarmot/yarnloom
>
> https://marketplace.visualstudio.com/items?itemName=TranquilMarmot.yarn-spinner-loom
>
> Any .yarn files made with Yarn Editor work with YarnLoom and vice versa!
> It uses the same mechanisms for node colors, tags, and placement.
>
> Key differences:
>
>    - Common library of functions used for parsing/generating yarn files
>    that are shared between the editor and the extension
>    - The editor sends much more specific messages back to the extension
>    so that it can track changes for better undo/redo functionality
>       - This means node renames, additions, color changes, edits, and
>       everything else support native undo/redo
>       - See the YarnLoom CONTRIBUTING.md
>       <https://github.com/TranquilMarmot/YarnLoom/blob/main/CONTRIBUTING.md#communication-between-loom-extension-and-loom-editor>
>       doc for more info
>    - Uses VSCode supplied CSS variables for colors so that it matches the
>    user's VSCode theme
>    - Uses codicons <https://github.com/microsoft/vscode-codicons> for
>    icons so they match VSCode
>    - Additional features
>       - More tag management tools
>       - Renaming nodes also renames links to those nodes
>       - Option for case sensitivity and regex when searching
>       - Better feedback around renaming/adding nodes with names that
>       already exist
>       - Uses D3 for drawing node relationships, resulting in much
>       smoother rendering
>    - Missing features (these may be added later)
>       - Preview using bondage.js
>       - Selecting and manipulating multiple nodes at once
>       - BBCode support
>       - JSON/XML/twee/twee2 import/export support (only .yarn files are
>       supported)
>    - Lots and lots of tests 😄
>    - 100% TypeScript 😉
>
> ------------------------------
>
> You're welcome to merge this and keep it!
>
> I'll be happy to answer any questions you have but will probably be
> putting more effort into maintaining YarnLoom.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/YarnSpinnerTool/VSCodeExtension/pull/1#issuecomment-667580861>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABRRWVK2S2INVWKW3KGIULTR6RY3PANCNFSM4NJHMX3Q>
> .
>

desplesda commented:
This is really nicely done, and I especially love your remix of Cecile’s logo!

Something I’d like to look at, sooner rather than later, is simplifying the Yarn editor ecosystem. Currently, there’s Yarn Editor (both as a web app and as a desktop app), this repo’s VS Code extension, the current VS Code extension, and in Unity there’s the Merino editor. Is consolidating any of these projects something you (and potentially other authors) would be interested in?

daviddq commented:
Super cool! I love the extension. Thanks a lot for your great work 👏 👏 👏 👏
